### PR TITLE
Fix issues with half/double and Sync Lock

### PR DESCRIPTION
**Pull request by ywwg**
One part of Sync Lock that isn't well-described by the code is the idea of which Syncable is responsible for initializing the Sync Leader values.  This is *not* the same as the leader! For instance, when you push and hold sync, that deck becomes the leader but it initializes based on the values of some other deck.  This concept was causing problems with half/double bpm calculation. The code is a little more explicit now, and doesn't just assume that the Sync Leader should be the source of sync params.

Also fixed some super edge cases

### CODE REVIEW

File: src/engine/sync/synccontrol.cpp [hunk 1]
```diff
@@ -289,7 +290,8 @@ void SyncControl::updateTargetBeatDistance() {
         targetDistance *= kBpmDouble;
     } else if (m_masterBpmAdjustFactor == kBpmHalve) {
         targetDistance *= kBpmHalve;
-        if (m_pBeatDistance->get() >= 0.5) {
+        // Our beat distance CO is still a buffer behind, so take the current value.
+        if (m_pBpmControl->getBeatDistance(getSampleOfTrack().current) >= 0.5) {
```
Comment (ywwg) [inline src/engine/sync/synccontrol.cpp:294]: very proud of this fix. It was hard to track down.

**Comment by ywwg** [comment]
p.s.: I really, really hate half-double syncing and wish I had never done it :P.

**Comment by ywwg** [comment]
The PR after this will rename master->leader! So let's get this in yes? :)

**Comment by Holzhaus** [comment]
Clazy checks failed

**Comment by ywwg** [comment]
> Clazy checks failed

fixed

File: src/test/enginesynctest.cpp [hunk 2]
```diff
@@ -1445,6 +1445,23 @@ TEST_F(EngineSyncTest, ZeroBPMRateAdjustIgnored) {
             ControlObject::getControl(ConfigKey(m_sGroup2, "rate"))->get());
 }
 
+TEST_F(EngineSyncTest, DISABLED_BeatDistanceBeforeStart) {
```
Comment (Holzhaus) [inline src/test/enginesynctest.cpp:1448]: Are you planning to fix this in this PR?

**Review by Holzhaus — COMMENTED** [review]
Thanks, I noticed no issues when testing this, but I can't really comment on the code changes because I'm not familiar enough with it. Would be good if someone else would have a look.

File: src/test/enginesynctest.cpp (hunk 2 above)
Comment (ywwg) [reply src/test/enginesynctest.cpp:1448]: no -- this is a much more complex issue and I'm not even sure what's going wrong.  This is a pretty rare circumstance, and shouldn't be a release-blocker, let alone a PR-blocker.

Comment (ywwg) [reply src/test/enginesynctest.cpp:1448]: https://bugs.launchpad.net/mixxx/+bug/1930143

**Review by daschuer — CHANGES_REQUESTED** [review]
I have a notable momentary pitch change in case of a implicit leader deck playing audible and a inaudible follower disables sync.  

* Play track A 
* Enable Sync A
* Move x-fader to A 
* Play track B 
* Enable Sync B
* A is implicit Leader
* Nudge Track A
* Disable Sync B 
* Hear a tempo dip in Track A -> wrong!

This can be noticed the best if Track A is a 440 Hz Sine wave analyzed with a const beat grid.

File: src/engine/sync/enginesync.cpp [hunk 3]
```diff
@@ -597,35 +603,68 @@ void EngineSync::setMasterBeatDistance(Syncable* pSource, double beatDistance) {
...
+    // as the Syncable setting the master parameters (here, pSource). Notify the proper Syncable
+    // so it can prepare itself.  (This is a hack to undo half/double math so that we initialize
+    // based on un-multiplied bpm values).
+    pSource->notifyMasterParamSource();
+
+    double beatDistance = pSource->getBeatDistance();
+    if (!pSource->isPlaying()) {
+        // If the params source is not playing, but other syncables are, then we are a stopped
+        // explicit Master and we should not initialize the beat distance.  Take it from the
+        // internal clock instead.
```
[the remark this answers is not in the document]
Comment (ywwg) [reply src/engine/sync/enginesync.cpp]: done

File: src/engine/sync/enginesync.cpp [hunk 4]
```diff
@@ -597,35 +603,68 @@ void EngineSync::setMasterBeatDistance(Syncable* pSource, double beatDistance) {
...
-        if (pSyncable == pSource || !pSyncable->isSynchronized()) {
+        if (!pSyncable->isSynchronized()) {
             continue;
         }
         pSyncable->setMasterParams(beatDistance, baseBpm, bpm);
     }
 }
 
-bool EngineSync::noPlayingFollowers() const {
+Syncable* EngineSync::getOnlyPlayingSyncable() const {
```
[the remark this answers is not in the document]
Comment (ywwg) [reply src/engine/sync/enginesync.cpp]: do you mean "lonely" ? :)  I tried "getUniquePlayingSyncedDeck"

**Comment by ywwg** [comment]
> I have a notable momentary pitch change in case of a implicit leader deck playing audible and a inaudible follower disables sync.

fixed.

**Comment by daschuer** [comment]
I can confirm the pitch dip issue is fixed with a nudged leader deck. 

Somehow the leader buttons are now broken. Is this a visual skin issue? The plain main branch is also affected. 

When I load a new track to a stopped explicit master deck the follower becomes high pitched. 
I don't think that this is a valid use case.

File: src/engine/sync/enginesync.cpp [hunk 5]
```diff
@@ -350,11 +354,15 @@ void EngineSync::notifyPlayingAudible(Syncable* pSyncable, bool playingAudible)
...
-    if (noPlayingFollowers() && m_pMasterSyncable) {
-        m_pMasterSyncable->notifyOnlyPlayingSyncable();
-        setMasterParams(m_pMasterSyncable);
+        setMasterParams(newMaster);
+    } else {
+        Syncable* pOnlyPlayer = getUniquePlayingSyncedDeck();
+        if (pOnlyPlayer) {
+            // Even if we didn't change master, if there is only one player (us), then we should
+            // reinit the beat distance.
+            pOnlyPlayer->notifyOnlyPlayingSyncable();
```
Comment (daschuer) [inline src/engine/sync/enginesync.cpp:363]: can you rename this to something like: 
notifyUniquePlaying() or such?

**Comment by ywwg** [comment]
> When I load a new track to a stopped explicit master deck the follower becomes high pitched.
> I don't think that this is a valid use case.

this is one of those edge cases.  It's an explicit leader, so it's supposed to be in charge of the rate.  Should we instead say that loading a new track causes it to relinquish explicit leader state? What's the right behavior?

**Comment by ronso0** [comment]
> Somehow the leader buttons are now broken. Is this a visual skin issue? The plain main branch is also affected.

fixed with 52786c9a76
we need to polish and add the Leader control to all skins anyway.

File: src/engine/sync/enginesync.cpp (hunk 5 above)
Comment (ywwg) [reply src/engine/sync/enginesync.cpp:363]: done

**Comment by ywwg** [comment]
Yes the leader button was always meant to be temp.  Is it ok if I leave it for now?

**Comment by ywwg** [comment]
hm, the leader button still seems broken to me

**Comment by ronso0** [comment]
just pick my commit for a temporary solution.
image

**Comment by ronso0** [comment]
it can't stay like that anyway since it screws up the deck layout

**Comment by daschuer** [comment]
> this is one of those edge cases. It's an explicit leader, so it's supposed to be in charge of the rate. Should we instead say that loading a new track causes it to relinquish explicit leader state? What's the right behavior?

On one hand it is the explicit leader, so the behavior can be explained. On the other hand I can't think of a use case where a playing track should change its tempo because a track is loaded into another deck. 
Do you have one? 

I think that should be fixed, either by relinquish explicit leader state or by sync the leader from a follower during load.

**Comment by ronso0** [comment]
> relinquish explicit leader state

requires the user to re-enable Leader.

> sync the leader from a follower during load

much more reasonable. Otherwise (avoid the follower to de/accelerate) you'd have to adjust the pitch fader in advance :\
Though, with an unwise track selection this may push the pitch fader out of bounds, thus the follower would de/accelerate anyway.
All in all, I'd rather expect this behaviour.

**Comment by ywwg** [comment]
> hm, the leader button still seems broken to me

fixed -- the qss got blown away in a bad merge I think

**Comment by ywwg** [comment]
I made the decks a few pixels taller to make room for the button.  Is that ok for now?

**Comment by ywwg** [comment]
Would it be ok if I fix the track load change in the next PR?  That's a separate bug from this one and this PR has been hanging around a while.  I have a good track record of fixing these things :)

**Comment by ywwg** [comment]
> sync the leader from a follower during load

This conflcts with our conclusion in  https://bugs.launchpad.net/mixxx/+bug/1808698.  There, we decided that "the second deck is still playing at unity -- it was the leader and it should not change speed just because it reloaded and the bpm changed."  The test we wrote has the other decks all changing speed to match the new value.

**Comment by ywwg** [comment]
I suppose the difference here is between when the beats are updated due to a track load (should resync to the playing bpm) vs beats are updated due to rescan (should keep the slider where it is).

**Comment by ywwg** [comment]
OK I have a new PR that fixes the track load behavior.  It is very far downstream of this PR so I'd really like to get this in so we can move ahead

**Comment by ywwg** [comment]
friendly ping -- I'd really like to start getting these in

**Comment by daschuer** [comment]
I have noticed an inconsistency: 
1:
* Play Deck A with Sync enabled (Implicit Leader)
* Nudge deck A
* Sync Deck B
* Play Deck B 
* -> Beat markers are in sync  

2:
* Play Deck A with Sync enabled (Implicit Leader)
* Sync Deck B (paused)
* Nudge deck A
* Play Deck B 
* -> Beat markers are out of sync  

I can imagine arguments for both behaviors, but both should behave the same. 
For now, I would prefer the first. Nudging is used to make two decks sound good, when playing together. 
When playing allown this feature is pointless.

**Comment by daschuer** [comment]
The behavior of nudging is IMHO correct, when playing with three decks. 
I can nudge the leader of two decks. When  I now add play a third synced deck it is in sync with the other follower, which is perfect. 

After rethinking the whole, my conclusion is: The current behavior is OK.

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
LGTM, Thank you.

**Comment by ywwg** [comment]
ok cool.  There's still a lot to fix with Sync Lock so we'll keep working on it

**Pull request merged**